### PR TITLE
fix: Set GCPMachinepoolMachine.Status.LatestModelApplied to false if some labels applied to node are not same as expected

### PR DESCRIPTION
**Pull request by tasdikrahman**
instancegroupinstances: set `gcpmachinepoolmachine.status.LatestModelApplied` to `false` if additionalLabels attached to node is not the same as ones set in `gcpmachinepool.spec.additionallabels` Which would mark it for cleanup in graceful way if `cluster.x-k8s.io/replicas-managed-by: "true"` is not present on the `MachinePool`, and the flow of https://github.com/newrelic-forks/cluster-api-provider-gcp/blob/v1.6.0-nr5/cloud/scope/machinepool.go#L278 would be called.  

 

 
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note

```

### CODE REVIEW

File: cloud/scope/machinepoolmachine.go [hunk 1]
```diff
@@ -318,15 +322,24 @@ func (m *MachinePoolMachineScope) HasLatestModelApplied(_ context.Context, insta
...
 	}
 	instanceImage := path.Base(diskImage.Path)
 
 	// Check if the image is the latest
-	if image == instanceImage {
+	// Check if the Labels applied to the instance are the latest as in the Instance Template.
+	// Remove the two keys of `capg-role` and `capg-cluster-<CLUSTER-NAME>` as they are added by default by CAPG
+	// and would be easier to compare the additional labels.
+	// ref: https://github.com/newrelic-forks/cluster-api-provider-gcp/blob/ef2e7f1e64ebeeb5389c446fe4cf89026fcb8a8a/cloud/services/compute/instances/reconcile_test.go#L244-L245
+	delete(instanceLabelsToCompare, "capg-role")
```
Comment (sachin-shankar) [inline cloud/scope/machinepoolmachine.go]: I wonder if removing the labels adds any value assuming if they are always present by default.

File: cloud/scope/machinepoolmachine.go [hunk 2]
```diff
@@ -302,7 +302,9 @@ func (m *MachinePoolMachineScope) ProviderID() string {
 }
 
 // HasLatestModelApplied checks if the latest model is applied to the GCPMachinePoolMachine.
-func (m *MachinePoolMachineScope) HasLatestModelApplied(_ context.Context, instance *compute.Disk) (bool, error) {
+func (m *MachinePoolMachineScope) HasLatestModelApplied(ctx context.Context, instance *compute.Disk, labels map[string]string) (bool, error) {
+	log := log.FromContext(ctx)
+
 	image := ""
```
Comment (sachin-shankar) [inline cloud/scope/machinepoolmachine.go]: I wonder if an explicit var declaration will help with the check failure.

File: cloud/scope/machinepoolmachine.go [hunk 3]
```diff
@@ -318,18 +320,46 @@ func (m *MachinePoolMachineScope) HasLatestModelApplied(_ context.Context, insta
...
 
-	// Check if the image is the latest
-	if image == instanceImage {
+	// Check if the image is the latest and additionalLabels
+	hasAdditionalLabelsDiff, err := m.hasAdditionalLabelsDiff(ctx, labels)
+	if err != nil {
+		log.Error(err, "Error checking the AdditonalLabels")
+		return false, err
+	}
+	if image == instanceImage && hasAdditionalLabelsDiff {
```
Comment (sachin-shankar) [inline cloud/scope/machinepoolmachine.go:330]: I think the name `hasAdditionalLabelsDiff` is a bit confusing here. Ideally it should mean `doesNotHaveAdditionalLabelsDiff` or `noAdditionalLabelsDiff` in the above context, correct?

File: cloud/scope/machinepoolmachine.go [hunk 4]
```diff
@@ -318,18 +320,46 @@ func (m *MachinePoolMachineScope) HasLatestModelApplied(_ context.Context, insta
...
 		return true, nil
 	}
 
+	log.Info("One of either Instance Image and AdditionalLabels are not the same as the ones specified in GCPMachinePool Spec, setting LatestModelApplied to false", "GCPMachinePoolMachine", m.GCPMachinePoolMachine)
 	return false, nil
 }
 
+// hasAdditionalLabelsDiff Checks if the Labels applied to the instance are the latest as in the Instance Template.
+// two keys of `capg-role` and `capg-cluster-<CLUSTER-NAME>` as they are added by default by CAPG
+// ref: https://github.com/newrelic-forks/cluster-api-provider-gcp/blob/ef2e7f1e64ebeeb5389c446fe4cf89026fcb8a8a/cloud/services/compute/instances/reconcile_test.go#L244-L24
```
Comment (sachin-shankar) [inline cloud/scope/machinepoolmachine.go:344]: Could you please add a one-liner to say we need to ignore the 2 labels?

Comment (tasdikrahman) [reply cloud/scope/machinepoolmachine.go:344]: just added a change here https://github.com/newrelic-forks/cluster-api-provider-gcp/pull/25/commits/761d5d79ea1bc944daf93a0a4fbab5af0249208b, please check

File: cloud/scope/machinepoolmachine.go (hunk 3 above)
Comment (tasdikrahman) [reply cloud/scope/machinepoolmachine.go:330]: just added a change here https://github.com/newrelic-forks/cluster-api-provider-gcp/pull/25/commits/761d5d79ea1bc944daf93a0a4fbab5af0249208b, please check

File: cloud/scope/machinepoolmachine.go (hunk 1 above)
Comment (tasdikrahman) [reply cloud/scope/machinepoolmachine.go]: resolving this comment, as this logic has been refactored.

File: cloud/scope/machinepoolmachine.go (hunk 2 above)
Comment (tasdikrahman) [reply cloud/scope/machinepoolmachine.go]: added a change, please check

**Comment by tasdikrahman** [comment]
```
$ make test                                                                                                                                                                                                                                   (wf2a/default)
KUBEBUILDER_ASSETS="/Users/trahman/Library/Application Support/io.kubebuilder.envtest/k8s/1.28.0-darwin-arm64" go test ./...
?   	sigs.k8s.io/cluster-api-provider-gcp	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/gcperrors	[no test files]
ok  	sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/cloud/providerid	(cached)
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/scope/strategies/machinepool_deployments	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/firewalls	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/instancegroupinstances	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/instancegroups	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/loadbalancers	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/networks	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/container/nodepools	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/shared	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/exp	[no test files]
ok  	sigs.k8s.io/cluster-api-provider-gcp/cloud/scope	1.597s
ok  	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/instances	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/compute/subnets	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/cloud/services/container/clusters	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/controllers	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/exp/api/v1beta1	(cached)
?   	sigs.k8s.io/cluster-api-provider-gcp/feature	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/hack/boilerplate/test	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/util/hash	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/util/location	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/util/resourceurl	[no test files]
?   	sigs.k8s.io/cluster-api-provider-gcp/version	[no test files]
ok  	sigs.k8s.io/cluster-api-provider-gcp/exp/controllers	1.421s
ok  	sigs.k8s.io/cluster-api-provider-gcp/util/processors	(cached)
ok  	sigs.k8s.io/cluster-api-provider-gcp/util/reconciler	(cached)
```

lint is passing. https://github.com/newrelic-forks/cluster-api-provider-gcp/actions/runs/9715896835/job/26818326647?pr=25

**Review by sachin-shankar — APPROVED** [review]

**Comment by mytunguyen** [comment]
@tasdikrahman go ahead and reopen this

**Pull request closed without merging**